python/tests/test_umap.py

```diff
@@ -0,0 +1,165 @@
+    supervised: bool,
+) -> float:
+    from cuml.manifold import UMAP
+
+    local_model = UMAP(n_neighbors=n_neighbors, random_state=42, init="random")
+    y_train = local_y if supervised else None
+    local_model.fit(local_X, y=y_train)
+
+    embedding = local_model.embedding_
+    return trustworthiness(local_X, embedding, n_neighbors=n_neighbors, batch_size=5000)
+
+
+def _spark_umap_trustworthiness(
+    local_X: np.ndarray,
+    local_y: np.ndarray,
+    n_neighbors: int,
+    supervised: bool,
+    n_parts: int,
+    sampling_ratio: float,
+    dtype: np.dtype,
+) -> float:
+    from spark_rapids_ml.umap import UMAP
+
+    local_model = UMAP(
+        n_neighbors=n_neighbors,
+        sample_fraction=sampling_ratio,
+        random_state=42,
+        init="random",
+        num_workers=n_parts,
+    )
+
+    with CleanSparkSession() as spark:
+        if supervised:
+            data_df, features_col, label_col = create_pyspark_dataframe(
+                spark, "array", dtype, local_X, local_y
+            )
+            assert label_col is not None
+            local_model.setLabelCol(label_col)
+        else:
+            data_df, features_col, _ = create_pyspark_dataframe(
+                spark, "array", dtype, local_X, None
+            )
+
+        local_model.setFeaturesCol(features_col)
+        distributed_model = local_model.fit(data_df)
+        embedding = cp.array(distributed_model.embedding_)
+
+    return trustworthiness(local_X, embedding, n_neighbors=n_neighbors, batch_size=5000)
+
+
+def _run_spark_test(
+    n_parts: int,
+    n_rows: int,
+    sampling_ratio: float,
+    supervised: bool,
+    dataset: str,
+    n_neighbors: int,
+    dtype: np.dtype,
+) -> bool:
+    local_X, local_y = _load_dataset(dataset, n_rows)
+
+    dist_umap = _spark_umap_trustworthiness(
+        local_X, local_y, n_neighbors, supervised, n_parts, sampling_ratio, dtype
+    )
+
+    loc_umap = _local_umap_trustworthiness(local_X, local_y, n_neighbors, supervised)
+
+    print("Local UMAP trustworthiness score : {:.2f}".format(loc_umap))
+    print("Spark UMAP trustworthiness score : {:.2f}".format(dist_umap))
+
+    trust_diff = loc_umap - dist_umap
+
+    return trust_diff <= 0.15
+
+
+@pytest.mark.parametrize("n_parts", [2, 9])
+@pytest.mark.parametrize("n_rows", [100, 500])
+@pytest.mark.parametrize("sampling_ratio", [1.0])
+@pytest.mark.parametrize("supervised", [True, False])
+@pytest.mark.parametrize("dataset", ["digits", "iris"])
+@pytest.mark.parametrize("n_neighbors", [10])
+@pytest.mark.parametrize("dtype", cuml_supported_data_types)
+def test_spark_umap(
+    n_parts: int,
+    n_rows: int,
+    sampling_ratio: float,
+    supervised: bool,
+    dataset: str,
+    n_neighbors: int,
+    dtype: np.dtype,
+) -> None:
+    result = _run_spark_test(
+        n_parts,
+        n_rows,
+        sampling_ratio,
+        supervised,
+        dataset,
+        n_neighbors,
+        dtype,
+    )
+
+    if not result:
```

Comment:
Could you please explain why we need to rerun it? Isn't the result determinate?

Reply:
This repetition is copied from cuml - you're right that the fit() should be deterministic, but this is primarily for testing transform(), which has stochasticity.